test/test_dataloader.py

```diff
@@ -2605,6 +2605,41 @@ def test_ind_worker_queue(self):
                 self._run_ind_worker_queue_test(batch_size=batch_size, num_workers=num_workers + 1)
 
 
+class TestMultiEpochDataset(IterableDataset):
+    def __init__(self, length):
+        self.length = length
+
+    def __iter__(self):
+        worker_info = torch.utils.data.get_worker_info()
+        assert worker_info is not None
+        worker_id = worker_info.id
+        for idx in range(self.length // worker_info.num_workers):
+            if worker_id == 0:
+                time.sleep(0.001)
+            yield worker_id
+
+    def __len__(self):
+        return self.length
+
+
+class TestMultiEpochDeterminism(TestCase):
+    def test_multi_epochs_deter(self):
+        num_workers = 2
+        batch_size = 10
+        num_epochs = 3
+
+        dataset = TestMultiEpochDataset(batch_size * num_workers)
+
+        for persistent in [True, False]:
+            loader = DataLoader(
+                dataset, batch_size=batch_size, num_workers=num_workers,
+                shuffle=False, persistent_workers=persistent, timeout=5
+            )
+            for ind in range(num_epochs):
+                for batch_idx, sample in enumerate(loader):
+                    self.assertEqual(sample.tolist(), [batch_idx % num_workers] * batch_size)
```

Comment:
super nit: Sorry for the confusion. I mean you can add this test under https://github.com/pytorch/pytorch/blob/5167e9d59d76d67f4a240f94600ce6bba2c70ed3/test/test_dataloader.py#L868
and use `self._get_data_loader` to get the dataloader.
`TestDataLoader` is inherited by `TestDataLoaderPersistentWorkers` to carry out the same test with `persistent_worker=True`. https://github.com/pytorch/pytorch/blob/5167e9d59d76d67f4a240f94600ce6bba2c70ed3/test/test_dataloader.py#L2334

Reply:
sure i can move it under that test.

Reply:
@ejguan with the sleep removed the test will randomly pass over 50% of the time for me.  I remove the sleep line, but just a heads up that it can lead to passing tests.

Hopefully I got the other changes you requested too.  

Thanks!